### ISSUE TITLE
Failing to import numpy.core.multiarray

### ISSUE THREAD

**Issue by landersson** [PDAL/PDAL#1010]
Hi,

I'm trying to use the predicate filter and I'm running into python module import issues... the command line:

pdal translate -i in.las -o out.las -f predicate --filters.predicate.script="$PWD/good.py" --filters.predicate.function=yow2 --filters.predicate.module=good

good.py is taken from the predicate filter unit tests:
```
import numpy as np
def yow2(ins,outs):
    Y = ins['Y']
    Mask = np.greater(Y, 1.0)
     #print Mask\n"
    outs['Mask'] = Mask
    return True
```
When running the command above, I get:

PDAL: ImportError('numpy.core.multiarray failed to import',)

It's failing when calling Py_CompileString() in Invocation.cpp::compile()

Running python_predicate_test from the build tree works fine... numpy is obviously installed, and running "import numpy.core.multiarray" in a python REPL seems to work fine too.

I've spent some time trying to debug this but I'm running out of ideas... I'm on Ubuntu 14.04 and using pdal/master.

**Comment by landersson** [comment]
Here's an strace from the command line above trying to access multiarray.so:

```
open("/usr/lib/python2.7/dist-packages/numpy/core/multiarray.so", O_RDONLY) = 12
fstat(12, {st_mode=S_IFREG|0644, st_size=1355072, ...}) = 0
open("/usr/lib/python2.7/dist-packages/numpy/core/multiarray.so", O_RDONLY|O_CLOEXEC) = 13
read(13, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\300\266\1\0\0\0\0\0"..., 832) = 832
fstat(13, {st_mode=S_IFREG|0644, st_size=1355072, ...}) = 0
mmap(NULL, 3457448, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 13, 0) = 0x7f37c6e81000
mprotect(0x7f37c6fbe000, 2093056, PROT_NONE) = 0
mmap(0x7f37c71bd000, 61440, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 13, 0x13c000) = 0x7f37c71bd000
mmap(0x7f37c71cc000, 4520, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f37c71cc000
close(13)                               = 0
munmap(0x7f37c6e81000, 3457448)         = 0
close(12)                               = 0
close(11)                               = 0
close(10)                               = 0
close(9)                                = 0
close(8)                                = 0
close(5)                                = 0
write(2, "ImportError", 11ImportError)             = 11
write(2, ": ", 2: )                       = 2
write(2, "numpy.core.multiarray failed to "..., 38numpy.core.multiarray failed to import) = 38
```
Doesn't give me a lot of clues though...

**Comment by abellgithub** [comment]
I can recreate this.  I'll look into it.

On Wed, Oct 21, 2015 at 3:59 AM, landersson <notifications@github.com>
wrote:

> Here's an strace from the command line above trying to access
> multiarray.so:
>
> open("/usr/lib/python2.7/dist-packages/numpy/core/multiarray.so", O_RDONLY) = 12
> fstat(12, {st_mode=S_IFREG|0644, st_size=1355072, ...}) = 0
> open("/usr/lib/python2.7/dist-packages/numpy/core/multiarray.so", O_RDONLY|O_CLOEXEC) = 13
> read(13, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\300\266\1\0\0\0\0\0"..., 832) = 832
> fstat(13, {st_mode=S_IFREG|0644, st_size=1355072, ...}) = 0
> mmap(NULL, 3457448, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 13, 0) = 0x7f37c6e81000
> mprotect(0x7f37c6fbe000, 2093056, PROT_NONE) = 0
> mmap(0x7f37c71bd000, 61440, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 13, 0x13c000) = 0x7f37c71bd000
> mmap(0x7f37c71cc000, 4520, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f37c71cc000
> close(13)                               = 0
> munmap(0x7f37c6e81000, 3457448)         = 0
> close(12)                               = 0
> close(11)                               = 0
> close(10)                               = 0
> close(9)                                = 0
> close(8)                                = 0
> close(5)                                = 0
> write(2, "ImportError", 11ImportError)             = 11
> write(2, ": ", 2: )                       = 2
> write(2, "numpy.core.multiarray failed to "..., 38numpy.core.multiarray failed to import) = 38
>
> Doesn't give me a lot of clues though...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PDAL/PDAL/issues/1010#issuecomment-149826490>.
>



-- 
Andrew Bell
andrew.bell.ia@gmail.com

**Comment by hobu** [comment]
My hunch is import_array being called before PyInitialize in some way.

**Comment by neteler** [comment]
We have the same issue - any patch we could test?

**Comment by hobu** [comment]
I haven't been able to reliably make this happen. Do you have a scenario that does, @abellgithub?

**Comment by abellgithub** [comment]
Yes.  Was thinking about kind of reverting stuff to where it was before we added the python -> pdal link just to get people going, and then try again.  I know things worked then.  Thoughts?

**Comment by hobu** [comment]
@abellgithub ok.

**Comment by abellgithub** [comment]
The problem is best described by this long-running Python bug report:

http://bugs.python.org/issue1454844

Since it seems like there is no interest in linking extension modules against libpython by the powers that be, we're need to bring the symbols in explicitly so they can be found when an extension is loaded.

**Comment by neteler** [comment]
I have locally patched my PDAL 1.0.1 (Scientific Linux 64bit, Python 2.7.5, numpy-1.7.1-10.el7.x86_64), the error still occurs.

pdal pipeline -v 1 -i /home/neteler/software/pdal_numpy_test.xml
ImportError: numpy.core.multiarray failed to import
PDAL: ImportError('numpy.core.multiarray failed to import',)

Anything I'm missing?

**Comment by abellgithub** [comment]
I need to know what compiler/linker you're using.  A man page for your dlopen() would also be helpful.

**Comment by neteler** [comment]
The compiler: gcc version 4.8.3 20140911 (Red Hat 4.8.3-9) (GCC) 
(it calls /usr/lib64/ccache/c++ i.e. ccache-3.1.9-3.el7.x86_64 : C/C++ compiler cache)

man dlopen attached:
[dlopen_man.txt](https://github.com/PDAL/PDAL/files/38981/dlopen_man.txt)

thanks!

**Comment by neteler** [comment]
Just FYI - I also tried with master and the same issue happens.

**Comment by abellgithub** [comment]
This should all work.  I'm thinking you're not locating the files you built.


Check:
1) The dates/times on libpdal_plang to verify it was built
2) Rebuild with the code from this morning.  If the proper code is getting executed, it should print "Force load python." on the command line.
3) Add LD_DEBUG=files before your pipeline command:
```
$  LD_DEBUG=files pdal pipeline mypipeline.xml
```
Search for libpdal_plang in the output to make sure it matches the library that was built.

**Comment by neteler** [comment]
I'm afraid that it uses the newly compiled library:

`[neteler@giscluster software]$ LD_DEBUG=files pdal pipeline /home/lucadelu/tmpmqtc07 2>&1 | grep 'numpy\|libpdal_plang'
      6743:	file=libpdal_plang.so.1 [0];  needed by /usr/lib64/libpdal_plugin_filter_programmable.so [0]
      6743:	file=libpdal_plang.so.1 [0];  generating link map
      6743:	calling init: /lib64/libpdal_plang.so.1
      6743:	file=/usr/lib64/python2.7/site-packages/numpy/core/multiarray.so [0];  dynamically loaded by /lib64/libpython2.7.so.1.0 [0]
      6743:	file=/usr/lib64/python2.7/site-packages/numpy/core/multiarray.so [0];  generating link map
      6743:	calling init: /usr/lib64/python2.7/site-packages/numpy/core/multiarray.so
      6743:	opening file=/usr/lib64/python2.7/site-packages/numpy/core/multiarray.so [0]; direct_opencount=1
      6743:	file=/usr/lib64/python2.7/site-packages/numpy/core/umath.so [0];  dynamically loaded by /lib64/libpython2.7.so.1.0 [0]
      6743:	file=/usr/lib64/python2.7/site-packages/numpy/core/umath.so [0];  generating link map
      6743:	calling init: /usr/lib64/python2.7/site-packages/numpy/core/umath.so
      6743:	opening file=/usr/lib64/python2.7/site-packages/numpy/core/umath.so [0]; direct_opencount=1
      6743:	file=/usr/lib64/python2.7/site-packages/numpy/core/_dotblas.so [0];  dynamically loaded by /lib64/libpython2.7.so.1.0 [0]
      6743:	file=/usr/lib64/python2.7/site-packages/numpy/core/_dotblas.so [0];  generating link map
      6743:	calling init: /usr/lib64/python2.7/site-packages/numpy/core/_dotblas.so
      6743:	opening file=/usr/lib64/python2.7/site-packages/numpy/core/_dotblas.so [0]; direct_opencount=1
ImportError: numpy.core.multiarray failed to import
PDAL: ImportError('numpy.core.multiarray failed to import',)
      6743:	calling fini: /lib64/libpdal_plang.so.1 [0]
      6743:	calling fini: /usr/lib64/python2.7/site-packages/numpy/core/multiarray.so [0]
      6743:	calling fini: /usr/lib64/python2.7/site-packages/numpy/core/umath.so [0]
      6743:	calling fini: /usr/lib64/python2.7/site-packages/numpy/core/_dotblas.so [0]
[neteler@giscluster software]$ 
[neteler@giscluster software]$ 
[neteler@giscluster software]$ ls -la /lib64/libpdal_plang.so.1*
lrwxrwxrwx 1 root root     22 Nov 19 15:39 /lib64/libpdal_plang.so.1 -> libpdal_plang.so.1.0.0*
-rwxr-xr-x 1 root root 407760 Nov 19 15:27 /lib64/libpdal_plang.so.1.0.0*
[neteler@giscluster software]$ 
[neteler@giscluster software]$ rpm -qf /lib64/libpdal_plang.so.1.0.0
PDAL-libs-1.0.2-2.el7.centos.x86_64`

**Comment by abellgithub** [comment]
I'm still suspicious since I'm not seeing the output that I asked about ("Force load python").  Rather than build/install an RPM, how about just running from the install location.  Go to the cmake build directory and execute bin/pdal pipeline <your pipeline>.  If you have PDAL_DRIVER_PATH set, unset it.

**Comment by landersson** [comment]
This seems to solve the problem for me. Thanks!

**Comment by neteler** [comment]
I repackaged with current master (I prefer to not clutter the system with individually installed package files): Now the "Force load python." comes up. Yet no success:

LD_DEBUG=files pdal pipeline /home/lucadelu/tmpmqtc07 2>&1 | tee log
     16313:	file=libpdal_base.so.1 [0];  needed by pdal [0]
     16313:	file=libpdal_base.so.1 [0];  generating link map
     16313:	  dynamic: 0x00007fa8ee927df8  base: 0x00007fa8ee2f0000   size: 0x000000000064f738
     16313:	    entry: 0x00007fa8ee50fd70  phdr: 0x00007fa8ee2f0040  phnum:                  7
     16313:	
     16313:	
     16313:	file=liblaszip.so.6 [0];  needed by pdal [0]
     16313:	file=liblaszip.so.6 [0];  generating link map
[...]
     16313:	file=libpython2.7.so.1.0 [0];  needed by /usr/lib64/libpdal_plugin_filter_programmable.so [0]
     16313:	file=libpython2.7.so.1.0 [0];  generating link map
     16313:	  dynamic: 0x00007fa8dd9f26b0  base: 0x00007fa8dd67a000   size: 0x00000000003c5608
     16313:	    entry: 0x00007fa8dd6b9370  phdr: 0x00007fa8dd67a040  phnum:                  7
     16313:	
     16313:	
     16313:	file=libutil.so.1 [0];  needed by /lib64/libpython2.7.so.1.0 [0]
     16313:	file=libutil.so.1 [0];  generating link map
     16313:	  dynamic: 0x00007fa8dd678dc0  base: 0x00007fa8dd477000   size: 0x0000000000202110
     16313:	    entry: 0x00007fa8dd477f10  phdr: 0x00007fa8dd477040  phnum:                  9
     16313:	
     16313:	
     16313:	calling init: /lib64/libutil.so.1
     16313:	
     16313:	
     16313:	calling init: /lib64/libpython2.7.so.1.0
     16313:	
     16313:	
     16313:	calling init: /lib64/libpdal_plang.so.1
     16313:	
Force load python.
     16313:	opening file= [0]; direct_opencount=2
     16313:	
     16313:	
     16313:	calling init: /usr/lib64/libpdal_plugin_filter_programmable.so
     16313:	
     16313:	opening file=/usr/lib64/libpdal_plugin_filter_programmable.so [0]; direct_opencount=1
     16313:	
     16313:	
     16313:	file=/usr/lib64/libpdal_plugin_filter_predicate.so [0];  dynamically loaded by /lib64/libpdal_base.so.1 [0]
     16313:	file=/usr/lib64/libpdal_plugin_filter_predicate.so [0];  generating link map
     16313:	  dynamic: 0x00007fa8dd475ba0  base: 0x00007fa8dd253000   size: 0x0000000000223d80
     16313:	    entry: 0x00007fa8dd268e80  phdr: 0x00007fa8dd253040  phnum:                  7
[...]
     16313:	calling init: /usr/lib64/python2.7/site-packages/numpy/core/_dotblas.so
     16313:	
     16313:	opening file=/usr/lib64/python2.7/site-packages/numpy/core/_dotblas.so [0]; direct_opencount=1
     16313:	
     16313:	
     16313:	file=/usr/lib64/python2.7/lib-dynload/cPickle.so [0];  dynamically loaded by /lib64/libpython2.7.so.1.0 [0]
     16313:	file=/usr/lib64/python2.7/lib-dynload/cPickle.so [0];  generating link map
     16313:	  dynamic: 0x00007fa8b86e9d28  base: 0x00007fa8b84d7000   size: 0x0000000000213de0
     16313:	    entry: 0x00007fa8b84daa30  phdr: 0x00007fa8b84d7040  phnum:                  7
     16313:	
     16313:	/usr/lib64/python2.7/lib-dynload/cPickle.so: error: symbol lookup error: undefined symbol: _Py_ZeroStruct (fatal)
     16313:	
     16313:	file=/usr/lib64/python2.7/lib-dynload/cPickle.so [0];  destroying link map
ImportError: numpy.core.multiarray failed to import
PDAL: ImportError('numpy.core.multiarray failed to import',)

     16313:	
     16313:	calling fini: /usr/lib64/libpdal_plugin_filter_predicate.so [0]
     16313:	
     16313:	
     16313:	file=/usr/lib64/libpdal_plugin_filter_predicate.so [0];  destroying link map
     16313:	
     16313:	calling fini: pdal [0]
     16313:	
     16313:	
     16313:	calling fini: /usr/lib64/libpdal_plugin_filter_programmable.so [0]
...


echo $PDAL_DRIVER_PATH
--> not set.

ldd /usr/lib64/python2.7/lib-dynload/cPickle.so 
	linux-vdso.so.1 =>  (0x00007fff5b588000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fd36be31000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fd36ba6f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fd36c284000)

nm -D /usr/lib64/python2.7/lib-dynload/cPickle.so | grep _Py_ZeroStruct
                 U _Py_ZeroStruct

Hope we can solve it.

**Comment by abellgithub** [comment]
Does it work with a test pipeline/python file as provided with the PDAL repository?

**Comment by neteler** [comment]
Yes: I run a few of them successfully, great. So the issue must now be in our own pipeline file, right?

**Comment by abellgithub** [comment]
Not sure of the problem.  I'm happy to help work it out.  Send me an email with your pipeline/python script.  andrew.bell.ia@gmail.com.

Closing the existing ticket.

**Comment by lucadelu** [comment]
In some special case this issue is still valid, @abellgithub has data to reproduce the error.

**Comment by lucadelu** [comment]
Now it is fixed also for my special case. Thanks a lot @abellgithub

**Comment by neteler** [comment]
On Apr 7, 2016 5:27 PM, "Luca Delucchi" <notifications@github.com> wrote:
>
> Now it is fixed also for my special case. Thanks a lot @abellgithub

Great! Into which release version will the fix be included?

**Comment by abellgithub** [comment]
It's in version 1.2.  The problem was a build issue.  Running cmake twice before compiling would also fix the issue in version 1.1, I think.  Run cmake once.  Make sure python is turned on.  Run cmake again and then run make/ninja.  In 1.2, it works the first time :)

**Comment by nickponline** [comment]
What need to be updated in https://github.com/PDAL/PDAL/blob/master/scripts/docker/1.6-maintenance/ubuntu/Dockerfile to make this work?
